Core/clim-basic/ports.lisp

```diff
@@ -171,107 +176,183 @@
+;;; sheets which doesn't have a mirror and that the sheet grabbing and
+;;; pressing is implemented locally. -- jd 2019-08-21
 (defmethod distribute-event ((port basic-port) (event pointer-event))
-  (when-let ((sheet (port-pointer-sheet port)))
-    (if (eq sheet (event-sheet event))
-        (call-next-method)
-        ;; events are immutable (explicitly stated in the spec) - that's why we
-        ;; need to make an event copy for single-mirrored sheets - event-sheet
-        ;; is not the same as sheet we want to distribute event to.
-        (let ((new-event (shallow-copy-object event)))
-          (setf (slot-value new-event 'sheet) sheet)
-          (unless (eq (sheet-mirrored-ancestor sheet)
-                      (sheet-mirrored-ancestor (event-sheet event)))
-            (multiple-value-bind (cx cy)
-                (untransform-position (sheet-delta-transformation (sheet-mirrored-ancestor sheet) nil)
-                                      (slot-value new-event 'graft-x)
-                                      (slot-value new-event 'graft-y))
-              (setf (slot-value new-event 'x) cx
-                    (slot-value new-event 'y) cy)))
-          (dispatch-event sheet new-event)))))
+  ;; When we receive pointer event we need to take into account
+  ;; unmirrored sheets and grabbed/pressed sheets.
+  ;;
+  ;; - Grabbed sheet steals all pointer events (non-local exit)
+  ;; - Pressed sheet receives replicated motion events
+  ;; - Pressing/releasing the button assigns pressed-sheet
+  ;; - Pressing the button sends the focus event
+  ;; - Pointer motion may result in synthetized boundary events
+  ;; - Events are delivered to the innermost child of the sheet
+  (when-let ((grabbed-sheet (port-grabbed-sheet port)))
+    (return-from distribute-event
+      (unless (typep event 'pointer-boundary-event)
+        (dispatch-event-copy grabbed-sheet event))))
+  ;; Synthetize boundary events and update the port-pointer-sheet.
+  (synthetize-boundary-events port event)
+  (let ((pressed-sheet (port-pressed-sheet port))
+        (new-pointer-sheet (port-pointer-sheet port)))
+    ;; Set the pointer cursor.
+    (when-let ((cursor-sheet (or pressed-sheet new-pointer-sheet)))
+      (let ((old-pointer-cursor (port-lookup-current-pointer-cursor port (event-sheet event)))
+            (new-pointer-cursor (sheet-pointer-cursor cursor-sheet)))
+        (unless (eql old-pointer-cursor new-pointer-cursor)
+          (set-sheet-pointer-cursor port (event-sheet event) new-pointer-cursor))))
+    ;; Handle some events specially.
+    (typecase event
+      ;; Pressing pointer button over a sheet makes a sheet
+      ;; pressed. Pressed sheet is assigned only when there is
+      ;; currently none.
+      (pointer-button-press-event
+       (when (null pressed-sheet)
+         (setf (port-pressed-sheet port) new-pointer-sheet))
+       (dispatch-event-copy new-pointer-sheet event 'window-manager-focus-event))
```

Comment:
Is `window-manager-focus-event` needed for this pull request? If not, please move it to the one concerned with input focus changes.

Reply:
fixed